back/lib/tasks/single_use/20240910_migrate_initiatives_to_proposals.rake

```diff
@@ -0,0 +1,811 @@
+      )
+    end
+  end
+end
+
+def rake_20240910_proposal_attributes(initiative, project)
+  proposals_phase = project.phases.first
+  {
+    slug: SlugService.new.generate_slug(Idea.new, initiative.slug),
+    title_multiloc: initiative.title_multiloc,
+    body_multiloc: initiative.body_multiloc,
+    author_id: initiative.author_id,
+    created_at: initiative.created_at,
+    updated_at: initiative.updated_at,
+    location_point_geojson: initiative.location_point_geojson,
+    location_description: initiative.location_description,
+    assignee_id: initiative.assignee_id,
+    assigned_at: initiative.assigned_at,
+    author_hash: initiative.author_hash,
+    anonymous: initiative.anonymous,
+    submitted_at: initiative.published_at || initiative.created_at,
+    project: project,
+    creation_phase: proposals_phase,
+    phases: [proposals_phase]
+  }
+end
+
+def rake_20240910_assign_idea_status(proposal, initiative, reporter)
+  code = initiative.initiative_status&.code
+  status = rake_20240910_to_idea_status(initiative.initiative_status)
+  if status
+    proposal.idea_status = status
+    true
+  elsif %w[review_pending changes_requested].include?(code)
+    proposal.idea_status = IdeaStatus.find_by(code: 'prescreening', participation_method: 'proposals')
+    true
+  else
+    reporter.add_error(
+      "No matching input status found for code #{code}",
+      context: { tenant: AppConfiguration.instance.host, proposal: proposal.slug, input_status_code: code }
+    )
+    false
+  end
+end
+
+def rake_20240910_to_idea_status(initiative_status)
+  IdeaStatus.find_by(code: initiative_status.code, participation_method: 'proposals')
+end
+
+def rake_20240910_assign_publication(proposal, initiative)
+  if proposal.idea_status.public_post?
+    proposal.publication_status = 'published'
+    proposal.published_at = initiative.proposed_at
+  else
+    proposal.publication_status = 'submitted'
+    proposal.submitted_at = initiative.created_at
+  end
+end
+
+def rake_20240910_migrate_status_changes(proposal, initiative, reporter)
+  previous_status_id = nil
+  initiative.initiative_status_changes.order(:created_at).each do |change|
+    next_status_id = rake_20240910_to_idea_status(change.initiative_status)&.id
+    activity = Activity.new(
+      item: proposal,
+      action: 'changed_status',
+      user: change.user,
+      acted_at: change.created_at,
+      payload: { change: [previous_status_id, next_status_id] }
+    )
+    previous_status_id = next_status_id
+    if !activity.save
+      reporter.add_error(
+        activity.errors.details,
+        context: { tenant: AppConfiguration.instance.host, proposal: proposal.slug, status_change: change.id }
+      )
+    end
+  end
+end
+
+def rake_20240910_migrate_images_files(proposal, initiative, reporter)
+  initiative.initiative_images.order(:ordering).each do |old_image|
+    new_image = IdeaImage.new(
```

Comment:
For all of these, should we set created_at and updated_at too? I think we probably should, as we're doing it for the initatives themselves as well.

Reply:
Will make the change :)